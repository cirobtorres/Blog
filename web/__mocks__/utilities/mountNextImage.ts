const nextOptimizedImage = (filePath: string) => {
  if (filePath) {
    const filename = filePath.substring(
      filePath.lastIndexOf("/") + 1,
      filePath.lastIndexOf(".")
    );
    const extension = filePath.substring(filePath.lastIndexOf(".") + 1);
    const path = filePath.substring(0, filePath.lastIndexOf("/") + 1);
    const URL = `${path}${filename}.${extension}`;
    const encodedURL = encodeURIComponent(URL);
    const width = "w=3840";
    const quality = "q=75";
    const nextOptimizedImage = `/_next/image?url=${encodedURL}&${width}&${quality}`;
    return nextOptimizedImage;
  }
  return null;
};

interface ToEncodeURIAttributes {
  imageFilename: string;
  imageFileExtension: string;
  path: string;
  width: number;
}

function mountLocalImage({
  imageFilename,
  imageFileExtension,
  path,
  width,
}: ToEncodeURIAttributes) {
  // Mock local images
  const quality = 75;
  const URL = `${path}${imageFilename}.${imageFileExtension}`;
  const encodedURL = encodeURIComponent(URL);
  const nextOptimizedImage = `/_next/image?url=${encodedURL}&w=${width}&q=${quality}`;
  return nextOptimizedImage;
}

function requestBackEndImage(
  filePath?: string,
  width: number = 1920,
  options?: {
    PROTOCOL?: string;
    IP?: string;
    PORT?: string;
  }
) {
  // Mock images managed by Strapi
  const quality = 75;

  if (filePath) {
    const filename = filePath.substring(
      filePath.lastIndexOf("/") + 1,
      filePath.lastIndexOf(".")
    );
    const extension = filePath.substring(filePath.lastIndexOf(".") + 1);
    const path = filePath.substring(0, filePath.lastIndexOf("/") + 1);

    const protocol = options ? options.PROTOCOL : "http";
    const ip = options ? options.IP : "127.0.0.1";
    const port = options ? options.PORT : "1337";
    const backendURL = `${protocol}://${ip}:${port}`;

    const URL = `${backendURL}${path}${filename}.${extension}`;

    const encodedURL = encodeURIComponent(URL);
    const nextOptimizedImage = `/_next/image?url=${encodedURL}&w=${
      width * 2
    }&q=${quality}`;

    return nextOptimizedImage;
  }

  const URL = "/images/not-authenticated.png";
  const encodedURL = encodeURIComponent(URL);

  const nextOptimizedImage = `/_next/image?url=${encodedURL}&w=${
    width * 2
  }&q=${quality}`;

  return nextOptimizedImage;
}

export { nextOptimizedImage, mountLocalImage, requestBackEndImage };
