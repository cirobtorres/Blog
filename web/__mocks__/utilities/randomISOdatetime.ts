import { faker } from "@faker-js/faker";

export function generateRandomCreatedAt() {
  const createdAt = faker.date.past().toISOString();
  const updatedAt = createdAt;
  const publishedAt = createdAt;
  return {
    createdAt,
    updatedAt,
    publishedAt,
  };
}
