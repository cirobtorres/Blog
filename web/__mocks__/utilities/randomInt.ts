import { faker } from "@faker-js/faker";

export const randomInt = (min: number, max: number) =>
  faker.number.int({ min, max });
