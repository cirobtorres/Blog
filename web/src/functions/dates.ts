function formatDateToCustomFormat(ISOdate: string) {
  const date = new Date(ISOdate);

  const months = [
    "janeiro",
    "fevereiro",
    "março",
    "abril",
    "maio",
    "junho",
    "julho",
    "agosto",
    "setembro",
    "outubro",
    "novembro",
    "dezembro",
  ];

  const day = date.getDate();
  const month = months[date.getMonth()];
  const year = date.getFullYear();
  const hours = String(date.getHours()).padStart(2, "0");
  const minutes = String(date.getMinutes()).padStart(2, "0");

  return `${day} de ${month} de ${year}, às ${hours}:${minutes}`;
}

export { formatDateToCustomFormat };
