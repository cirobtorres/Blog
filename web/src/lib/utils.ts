import { remark } from "remark";
import html from "remark-html";
import { clsx, type ClassValue } from "clsx";
import { twMerge } from "tailwind-merge";

const cn = (...inputs: ClassValue[]) => {
  return twMerge(clsx(inputs));
};

const convertMarkdowToHtmlString = async (rawHtml: string | undefined) => {
  const processedContent = await remark().use(html).process(rawHtml);
  const stringHtml = processedContent.toString();
  return stringHtml;
};

export { cn, convertMarkdowToHtmlString };
