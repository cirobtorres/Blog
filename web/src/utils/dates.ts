function formatDateToCustomFormat(ISOdate: string) {
  const date = new Date(ISOdate);

  const months = [
    "janeiro",
    "fevereiro",
    "março",
    "abril",
    "maio",
    "junho",
    "julho",
    "agosto",
    "setembro",
    "outubro",
    "novembro",
    "dezembro",
  ];

  const day = date.getDate();
  const month = months[date.getMonth()];
  const year = date.getFullYear();
  const hours = String(date.getHours()).padStart(2, "0");
  const minutes = String(date.getMinutes()).padStart(2, "0");

  return `${day} de ${month} de ${year}, às ${hours}:${minutes}`;
}

function formatDateToYouTubeLikeFormat(ISOdate: string) {
  const date = new Date(ISOdate);
  const now = new Date();
  const diffInMs = now.getTime() - date.getTime();
  const diffInSeconds = Math.floor(diffInMs / 1000);
  const diffInMinutes = Math.floor(diffInSeconds / 60);
  const diffInHours = Math.floor(diffInMinutes / 60);
  const diffInDays = Math.floor(diffInHours / 24);
  const diffInMonths = Math.floor(diffInDays / 30);

  if (diffInSeconds < 60) {
    return `${diffInSeconds <= 1 ? "agora" : `há ${diffInSeconds} segundos`}`;
  }

  if (diffInMinutes < 60) {
    return `há ${diffInMinutes} ${diffInMinutes === 1 ? "minuto" : "minutos"}`;
  }

  if (diffInHours < 24) {
    return `há ${diffInHours} ${diffInHours === 1 ? "hora" : "horas"}`;
  }

  if (diffInDays < 30) {
    return `há ${diffInDays} ${diffInDays === 1 ? "dia" : "dias"}`;
  }

  return `há ${diffInMonths} ${diffInMonths === 1 ? "mês" : "meses"}`;
}

export { formatDateToCustomFormat, formatDateToYouTubeLikeFormat };
