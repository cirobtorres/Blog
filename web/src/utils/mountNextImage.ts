interface ToEncodeURIAttributes {
  imageFilename: string;
  imageFileExtension: string;
  path: string;
  width: number;
  quality: number;
}

function mountNextImage({
  imageFilename,
  imageFileExtension,
  path,
  width,
  quality,
}: ToEncodeURIAttributes) {
  const URL = `${path}${imageFilename}.${imageFileExtension}`;
  const encodedURL = encodeURIComponent(URL);
  const nextOptimizedImage = `/_next/image?url=${encodedURL}&w=${width}&q=${quality}`;
  return nextOptimizedImage;
}

function requestBackEndNextImage(
  imageFilename: string,
  imageFileExtension: string,
  path: string,
  width: number,
  quality: number,
  options?: {
    PROTOCOL?: string;
    IP?: string;
    PORT?: string;
  }
) {
  const URL = `${options ? options.PROTOCOL : "http"}://${
    options ? options.IP : "127.0.0.1"
  }:${
    options ? options.PORT : "1337"
  }${path}${imageFilename}.${imageFileExtension}`;
  const encodedURL = encodeURIComponent(URL);
  const nextOptimizedImage = `/_next/image?url=${encodedURL}&w=${width}&q=${quality}`;

  return nextOptimizedImage;
}

export { mountNextImage, requestBackEndNextImage };
