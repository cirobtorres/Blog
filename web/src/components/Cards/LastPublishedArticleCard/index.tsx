import Image from "next/image";
import Link from "next/link";
import { formatDateToCustomFormat } from "../../../utils/dates";

const LastPublishedArticleCard = ({ article }: { article: ArticleCard }) => {
  return (
    <div className="w-full py-4 bg-blog-background-2 mb-10 border-y border-blog-border">
      <div className="max-w-screen-2xl mx-auto flex items-center">
        <Link
          href={`artigos/${article.documentId}/${article.slug}`}
          className="w-full"
        >
          <div className="relative h-[500px]">
            <Image
              src={`http://127.0.0.1:1337${article.cover.url}`}
              alt={article.cover.alternativeText}
              fill
              sizes={
                `(max-width: ${article.cover.width}) 100vw`
                // + `, (max-width: ${article.cover.width / 2}) 50vw`
                // + `, 33vw`
              }
              className="absolute object-cover"
            />
            <div className="flex flex-col gap-2 w-full absolute bottom-0 px-20 max-[800px]:px-12 max-[500px]:px-4 py-4 backdrop-blur-xl backdrop-brightness-150">
              <p
                className={
                  "text-white font-extrabold" +
                  " text-4xl max-[800px]:text-3xl max-[500px]:text-2xl"
                }
              >
                {article.title}
              </p>
              <p
                className={
                  "text-white" +
                  " text-xl max-[800px]:text-lg max-[500px]:text-base" +
                  " max-h-[calc(1.75rem_*_3)] line-clamp-3" +
                  // " max-[800px]:max-h-[calc(1.75rem_*_3)] max-[800px]:line-clamp-3" +
                  " max-[500px]:max-h-[calc(1.5rem_*_2)] max-[500px]:line-clamp-2"
                }
              >
                {article.description}
              </p>
              <p>
                <small className="flex justify-center text-white text-xs">
                  <time>{formatDateToCustomFormat(article.createdAt)}</time>
                </small>
              </p>
            </div>
          </div>
        </Link>
      </div>
    </div>
  );
};

export default LastPublishedArticleCard;
