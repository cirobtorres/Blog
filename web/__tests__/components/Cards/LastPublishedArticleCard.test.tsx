import { faker } from "@faker-js/faker";
import { render, screen } from "@testing-library/react";
import {
  createArticleMock,
  mockArticle,
} from "../../../__mocks__/mockArticles";
import { requestBackEndImage } from "../../../__mocks__/utilities/mountNextImage";
import { formatDateToCustomFormat } from "../../../src/utils/dates";
import LastPublishedArticleCard from "../../../src/components/Cards/LastPublishedArticleCard";

faker.seed(5); // Snapshot

const snapshotArticle = createArticleMock();

describe("LastPublishedArticleCard", () => {
  beforeEach(() => {
    process.env.NEXT_PUBLIC_BACKEND_IP = "http://127.0.0.1:1337";
  });

  it("renders last published article card container component", () => {
    render(<LastPublishedArticleCard article={mockArticle} />);
    const container = screen.getByTestId(
      "last-published-article-card-container"
    );
    expect(container).toBeInTheDocument();
    expect(container).toHaveClass(
      "w-full py-4 mb-10 border-y border-blog-border bg-blog-background-2"
    );
  });

  it("renders last published article card link", () => {
    render(<LastPublishedArticleCard article={mockArticle} />);
    const container = screen.getByTestId("last-published-article-card-link");
    expect(container).toBeInTheDocument();
    expect(container).toHaveClass("w-full");
    expect(container).toHaveAttribute(
      "href",
      `artigos/${mockArticle.documentId}/${mockArticle.slug}`
    );
  });

  it("renders last published article card image", () => {
    render(<LastPublishedArticleCard article={mockArticle} />);
    const container = screen.getByTestId(
      "last-published-article-card-image-container"
    );
    expect(container).toBeInTheDocument();
    expect(container).toHaveClass("relative h-[500px]");
    const link = container.querySelector("img");
    expect(link).toBeInTheDocument();
    expect(link).toHaveClass("absolute object-cover");
    expect(link).toHaveAttribute(
      "src",
      requestBackEndImage(mockArticle.cover.url)
    );
    expect(link).toHaveAttribute("alt", mockArticle.cover.alternativeText);
    expect(link).toHaveAttribute("data-nimg", "fill");
    expect(link).toHaveAttribute(
      "sizes",
      `(max-width: ${mockArticle.cover.width}) 100vw`
    );
  });

  it("renders last published article card paragraph container", () => {
    render(<LastPublishedArticleCard article={mockArticle} />);
    const container = screen.getByTestId(
      "last-published-article-card-paragraph-container"
    );
    expect(container).toBeInTheDocument();
    expect(container).toHaveClass(
      "w-full absolute bottom-0 flex flex-col gap-2 px-20 max-[800px]:px-12 max-[500px]:px-4 py-4 backdrop-blur-xl backdrop-brightness-150"
    );
    const [p1, p2, p3] = container.querySelectorAll("p");
    expect(p1).toBeInTheDocument();
    expect(p1).toHaveTextContent(mockArticle.title);
    expect(p1).toHaveClass(
      "text-white font-extrabold text-4xl max-[800px]:text-3xl max-[500px]:text-2xl"
    );
    expect(p2).toBeInTheDocument();
    expect(p2).toHaveTextContent(mockArticle.description);
    expect(p2).toHaveClass(
      "text-white text-xl max-[800px]:text-lg max-[500px]:text-base max-h-[calc(1.75rem_*_3)] line-clamp-3 max-[500px]:max-h-[calc(1.5rem_*_2)] max-[500px]:line-clamp-2"
    );
    expect(p3).toBeInTheDocument();
    const small = p3.querySelector("small");
    expect(small).toBeInTheDocument();
    expect(small).toHaveClass("flex justify-center text-white text-xs");
    const time = small?.querySelector("time");
    expect(time).toBeInTheDocument();
    expect(time).toHaveTextContent(
      formatDateToCustomFormat(mockArticle.createdAt)
    );
  });

  it("matches the snapshot", () => {
    const { asFragment } = render(
      <LastPublishedArticleCard
        article={{
          ...snapshotArticle,
          createdAt: "2021-12-03T05:40:44.408Z",
          updatedAt: "2021-12-03T05:40:44.408Z",
          publishedAt: "2021-12-03T05:40:44.408Z",
        }}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
