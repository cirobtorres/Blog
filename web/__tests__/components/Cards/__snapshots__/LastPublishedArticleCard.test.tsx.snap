// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LastPublishedArticleCard matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="w-full py-4 mb-10 border-y border-blog-border bg-blog-background-2"
    data-testid="last-published-article-card-container"
    id="last-published-article-card-container"
  >
    <div
      class="max-w-screen-2xl mx-auto flex items-center"
    >
      <a
        class="w-full"
        data-testid="last-published-article-card-link"
        href="artigos/3d3e79c8-431b-472e-8464-a99444522ff3/subseco-tergum-acerbitas-nesciunt-ab-dolorum"
        id="last-published-article-card-link"
      >
        <div
          class="relative h-[500px]"
          data-testid="last-published-article-card-image-container"
          id="last-published-article-card-image-container"
        >
          <img
            alt="Molestias cognatus tremo soluta vos amita nobis demergo solio carbo."
            class="absolute object-cover"
            data-nimg="fill"
            decoding="async"
            loading="lazy"
            sizes="(max-width: 512) 100vw"
            src="/_next/image?url=http%3A%2F%2F127.0.0.1%3A1337%2Fthe%2Fup%2Fout%2Fsociable%2Fphooey_through_bare.png&w=3840&q=75"
            srcset="/_next/image?url=http%3A%2F%2F127.0.0.1%3A1337%2Fthe%2Fup%2Fout%2Fsociable%2Fphooey_through_bare.png&w=640&q=75 640w, /_next/image?url=http%3A%2F%2F127.0.0.1%3A1337%2Fthe%2Fup%2Fout%2Fsociable%2Fphooey_through_bare.png&w=750&q=75 750w, /_next/image?url=http%3A%2F%2F127.0.0.1%3A1337%2Fthe%2Fup%2Fout%2Fsociable%2Fphooey_through_bare.png&w=828&q=75 828w, /_next/image?url=http%3A%2F%2F127.0.0.1%3A1337%2Fthe%2Fup%2Fout%2Fsociable%2Fphooey_through_bare.png&w=1080&q=75 1080w, /_next/image?url=http%3A%2F%2F127.0.0.1%3A1337%2Fthe%2Fup%2Fout%2Fsociable%2Fphooey_through_bare.png&w=1200&q=75 1200w, /_next/image?url=http%3A%2F%2F127.0.0.1%3A1337%2Fthe%2Fup%2Fout%2Fsociable%2Fphooey_through_bare.png&w=1920&q=75 1920w, /_next/image?url=http%3A%2F%2F127.0.0.1%3A1337%2Fthe%2Fup%2Fout%2Fsociable%2Fphooey_through_bare.png&w=2048&q=75 2048w, /_next/image?url=http%3A%2F%2F127.0.0.1%3A1337%2Fthe%2Fup%2Fout%2Fsociable%2Fphooey_through_bare.png&w=3840&q=75 3840w"
            style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
          />
          <div
            class="w-full absolute bottom-0 flex flex-col gap-2 px-20 max-[800px]:px-12 max-[500px]:px-4 py-4 backdrop-blur-xl backdrop-brightness-150"
            data-testid="last-published-article-card-paragraph-container"
            id="last-published-article-card-paragraph-container"
          >
            <p
              class="text-white font-extrabold text-4xl max-[800px]:text-3xl max-[500px]:text-2xl"
            >
              Subseco tergum acerbitas nesciunt ab dolorum.
            </p>
            <p
              class="text-white text-xl max-[800px]:text-lg max-[500px]:text-base max-h-[calc(1.75rem_*_3)] line-clamp-3 max-[500px]:max-h-[calc(1.5rem_*_2)] max-[500px]:line-clamp-2"
            >
              Voro caute thorax usus vesco abbas delinquo vomica cubitum torqueo in tepidus deorsum adamo aiunt amoveo casso vobis ratione totam maxime rem tonsor.
            </p>
            <p>
              <small
                class="flex justify-center text-white text-xs"
              >
                <time>
                  3 de dezembro de 2021, Ã s 02:40
                </time>
              </small>
            </p>
          </div>
        </div>
      </a>
    </div>
  </div>
</DocumentFragment>
`;
